### PR TITLE
[PyTorch FE] Support det, svd, rank>2 linalg_norm and dynamic-kernel max_pool2d

### PR DESCRIPTION
**Pull request by evkotov**
### Details:
Enables direct ov.convert_model of the SAM-6D Pose Estimation Model, which
previously failed on a set of unsupported PyTorch frontend ops. All changes are
frontend-only (no core op or plugin changes).

Changes:

- linalg_norm: support rank > 2 without an explicit ord. ord=None now selects
  the Frobenius norm when reducing over two axes and the vector L2 norm
  otherwise; both reduce to sqrt(sum(x^2)) over the selected axes, so the
  computation is rank-agnostic. Removes the artificial rank>2 throw in norm.cpp.

- aten::det / aten::linalg_det: new translator (op/linalg_det.cpp), 3x3 cofactor
  expansion. Registered for both names on the TorchScript path.

- aten::svd: new translator (op/svd.cpp), one-sided Jacobi SVD for 3x3, fp32
  stable including rank-deficient Procrustes inputs.

- aten::max_pool2d with a dynamic kernel_size: when the kernel is a runtime
  value (e.g. kernel_size=[1, x.size(3)]), the static MaxPool path cannot be
  used since OV MaxPool takes the kernel as a constructor attribute. The static
  path is preserved behind a per-element constness probe; a non-constant kernel
  takes a values-only ReduceMax decomposition over the full-extent axes.
  return_indices and partial (non-full-extent) dynamic windows raise a clear
  conversion error.

Scope notes:
- det/svd are scoped to the 3x3 matrices SAM-6D uses. On the TorchScript path
  the decoder forces all dims dynamic, so a shared ensure_trailing_square helper
  (utils.cpp) inserts a runtime reshape guard to [n, n]: an identity for a real
  3x3 input, a loud runtime error for any other size instead of a silent wrong
  result.
- Coverage is the TorchScript path; FX/torch.compile registration is a follow-up
  (torch.export lowers det to aten::_linalg_det).

### Tickets:
 - 181666

### CODE REVIEW

File: src/frontends/pytorch/src/op/max_poolnd.cpp [hunk 1]
```diff
@@ -28,9 +30,132 @@ namespace pytorch {
...
+                                    "d with a non-constant kernel_size is only supported with ceil_mode=False.");
+    }
+
+    // Decide, per spatial axis, whether it is reduced (full-extent) or left untouched (window 1).
+    std::vector<int64_t> reduce_axes;
+    for (int i = 0; i < dims; ++i) {
+        const int64_t axis = static_cast<int64_t>(i) - dims;  // negative index of this spatial axis
+        if (!elem_is_const[i]) {
+            // Runtime kernel element: assume it spans the whole axis (global pool over that axis).
+            reduce_axes.push_back(axis);
```
Comment (mvafin) [inline src/frontends/pytorch/src/op/max_poolnd.cpp:143]: A non-constant kernel element is always assumed to be a full-extent global pool and lowered to ReduceMax, with no check that the runtime kernel equals the axis extent. A genuine dynamic strided pool (e.g. kernel_size=k, k < extent, default stride) passes all guards and silently produces [N,C,1,1] instead of the strided result. Scalar case (L229) collapses both axes. Can we add a runtime guard (kernel == axis extent) so this fails loudly instead?

File: src/frontends/pytorch/src/utils.cpp [hunk 2]
```diff
@@ -160,6 +160,43 @@ Output<Node> reshape_kernel_for_group(const NodeContext& context, const Output<N
...
+    // trailing two axes to a fixed [n, n] while preserving the batch axes.
+    auto shape = context.mark_node(std::make_shared<v3::ShapeOf>(x, element::i64));
+    auto start = context.mark_node(v0::Constant::create(element::i64, Shape{1}, {0}));
+    auto stop = context.mark_node(v0::Constant::create(element::i64, Shape{1}, {-2}));
+    auto step = context.mark_node(v0::Constant::create(element::i64, Shape{1}, {1}));
+    auto axis = context.mark_node(v0::Constant::create(element::i64, Shape{1}, {0}));
+    auto batch_shape = context.mark_node(std::make_shared<v8::Slice>(shape, start, stop, step, axis));
+    auto nn = context.mark_node(v0::Constant::create(element::i64, Shape{2}, {n, n}));
+    auto new_shape = context.mark_node(std::make_shared<v0::Concat>(OutputVector{batch_shape, nn}, 0));
+    return context.mark_node(std::make_shared<v1::Reshape>(x, new_shape, /*special_zero=*/false));
```
Comment (mvafin) [inline src/frontends/pytorch/src/utils.cpp]: det/svd are registered for the general ops but only 3×3 works. On the dynamic path a torch-valid 2×2/4×4 input gets a bare runtime Reshape mismatch instead of the clean "only 3×3 supported" message the static path gives. Could the dynamic path surface an equivalent op-labeled error?

File: src/frontends/pytorch/src/op/svd.cpp [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+        vq = nvq;
+    }
+
+    Output<Node> absval(const Output<Node>& x) {
+        // |x| via Select(x < 0, -x, x); avoids pulling in an Abs op and keeps full
+        // precision (no sqrt rounding).
+        auto neg = mul(x, cf(-1.0f));
+        return sel(m_ctx.mark_node(std::make_shared<v1::Less>(x, cf(0.0f))), neg, x);
+    }
+    Output<Node> signum(const Output<Node>& x) {
```
Comment (mvafin) [inline src/frontends/pytorch/src/op/svd.cpp:196]: These reimplement Abs/Sign via Less+Multiply(-1)+Select. The "avoids pulling in an Abs op / keeps full precision" comment is inaccurate — v0::Abs is an exact sign-bit clear with no rounding. Called per rotation, this adds ~100 ops. Suggest v0::Abs/v0::Sign and drop the comment.

File: src/frontends/pytorch/src/op/norm.cpp [hunk 4]
```diff
@@ -269,15 +269,28 @@ OutputVector translate_linalg_norm(const NodeContext& context) {
...
+    // Both branches reduce to sqrt(sum(x^2)) over the selected axes (x^2 == |x|^2 for
+    // the real inputs supported here), so the computation itself is rank-agnostic; only
+    // the branch selection below needs the static rank, for the dim=None case.
     if (context.input_is_none(1)) {
-        auto input_rank = x.get_partial_shape().rank();
-        if (input_rank.is_static() && input_rank.get_length() == 2) {
+        bool is_matrix_norm = false;
+        if (!context.input_is_none(2)) {
+            // explicit dim provided: matrix norm iff it lists exactly two axes
+            auto const_dim = context.const_input<std::vector<int64_t>>(2);
```
Comment (mvafin) [inline src/frontends/pytorch/src/op/norm.cpp]: The ord=None branch now calls const_input<...>(2) just to count axes, requiring dim to be foldable — a non-constant dim now aborts conversion where it previously worked. Rare in practice, but a fallback to the rank-based path would avoid the regression.

File: src/frontends/pytorch/src/op/svd.cpp [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+        return m_ctx.mark_node(std::make_shared<v8::Gather>(m, ci(k), ci_s(-1)));  // (...,3,1)
+    }
+    // Same as mcol but for the V columns of an identity-like start. We build the
+    // identity columns from constants broadcast to the batch of `like`.
+    Output<Node> ident_col(const Output<Node>& like, int64_t k) {
+        std::vector<float> e(3, 0.0f);
+        e[static_cast<size_t>(k)] = 1.0f;
+        auto col = m_ctx.mark_node(v0::Constant::create(m_et, Shape{3, 1}, e));
+        // Broadcast to batch via add with 0*like-column (keeps the (...,3,1) shape).
+        auto zero_like = mul(mcol(like, 0), cf(0.0f));
```
Comment (mvafin) [inline src/frontends/pytorch/src/op/svd.cpp:129]: These reimplement Abs/Sign via Less+Multiply(-1)+Select. The "avoids pulling in an Abs op / keeps full precision" comment is inaccurate — v0::Abs is an exact sign-bit clear with no rounding. Called per rotation, this adds ~100 ops. Suggest v0::Abs/v0::Sign and drop the comment.

**Review by mvafin — CHANGES_REQUESTED** [review]
You add support only for 3x3 det and svd versions of operations. Please make sure the operations with other dimentions will fail gracefully

File: src/frontends/pytorch/src/op/max_poolnd.cpp (hunk 1 above)
Comment (evkotov) [reply src/frontends/pytorch/src/op/max_poolnd.cpp:143]: Good catch. Added a per-reduced-axis runtime guard: before the ReduceMax, `guard_full_extent_axis` inserts a Reshape that pins the pooled axis to the runtime kernel value while copying the other axes from ShapeOf. When the extent equals the kernel it is an identity; a genuine strided pool (k < extent) now fails loudly at inference instead of silently returning [N,C,1,1]. The guard node is named `aten::max_poolNd/require_full_extent` so the CPU error is op-labeled. This also covers the scalar-kernel case (each axis is guarded against its own extent). Verified that a `kernel_size=[1, x.size(3)//2]` model raises `[CPU] Reshape node with name 'aten::max_pool2d/require_full_extent' ... conflicts with the reshape pattern`. Added a negative test `test_max_pool_dynamic_partial_kernel_fails_at_runtime`.

File: src/frontends/pytorch/src/utils.cpp (hunk 2 above)
Comment (evkotov) [reply src/frontends/pytorch/src/utils.cpp]: Done. The dynamic-path guard Reshape in `ensure_trailing_square` now gets an op-labeled friendly_name (`<op>/requires_NxN`), which the CPU plugin embeds in the runtime message. A dynamic 4x4 det now fails with `[CPU] Reshape node with name 'aten::det/linalg_det/requires_3x3' ... conflicts with the reshape pattern`, mirroring the clean static-shape conversion-time message.

File: src/frontends/pytorch/src/op/svd.cpp (hunk 3 above)
Comment (evkotov) [reply src/frontends/pytorch/src/op/svd.cpp:196]: Switched `absval` to v0::Abs and removed the inaccurate comment. For the sign I adopted v0::Sign but kept a small guard: `Select(Equal(x,0), +1, Sign(x))`. v0::Sign(0)=0 would set the Jacobi t=0 and skip the +-45 degree rotation exactly when two columns have equal norm (zeta = (beta-alpha)/(2*gamma) = 0 with gamma != 0) -- reachable in fp32 for duplicated/rank-deficient columns, which is the pose-estimation Procrustes regime. The guard keeps sign(0)=+1 so that rotation still fires. Confirmed the rank-deficient SVD/Procrustes tests still match torch (R to ~1.8e-6).

File: src/frontends/pytorch/src/op/svd.cpp (hunk 5 above)
Comment (evkotov) [reply src/frontends/pytorch/src/op/svd.cpp:129]: See the reply on the paired comment: `absval` is now v0::Abs and `signum` uses v0::Sign wrapped as `Select(Equal(x,0), +1, Sign(x))` to preserve sign(0)=+1 (otherwise the +-45 degree Jacobi rotation is skipped for equal-norm / rank-deficient columns). Inaccurate comment removed.

File: src/frontends/pytorch/src/op/norm.cpp (hunk 4 above)
Comment (evkotov) [reply src/frontends/pytorch/src/op/norm.cpp]: Fixed by removing the dependency entirely: for ord=None, Frobenius and vector L2 are both sqrt(sum(x^2)) over the selected axes, so the is_matrix_norm branch is unnecessary. The ord=None path now always calls `norm_vector(x, dim, 2, keepdim)` -- no const_input(2), no static-rank requirement, identical result.

File: tests/layer_tests/pytorch_tests/test_det.py [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+
+        return aten_det()
+
+    @pytest.mark.nightly
+    @pytest.mark.precommit
+    def test_static_4x4_fails_at_conversion(self, ie_device, precision, ir_version):
+        # Static trailing dims => the conversion-time check in ensure_trailing_square fires.
+        example = torch.randn(2, 4, 4, dtype=torch.float32)
+        scripted = torch.jit.trace(self._det(), example)
+        with pytest.raises(Exception):
```
[the remark this answers is not in the document]
Comment (evkotov) [reply tests/layer_tests/pytorch_tests/test_det.py]: Done for the two cases that actually raise it — `test_compute_uv_false_fails_at_conversion` and the max_pool `sliding_window_unsupported` test now assert `ov.frontend.OpConversionFailure` (both fail inside the translator's `PYTORCH_OP_CONVERSION_CHECK`).

The other blocks do not raise `OpConversionFailure`, so asserting it there would be incorrect: the static non-square det/svd guard trips **core** Reshape shape-inference during model build (`reshape_shape_inference.hpp`), and the dynamic / partial-kernel cases fail in the CPU plugin at inference — both are plain `RuntimeError`. For those I kept `pytest.raises(Exception)` but now assert the op-labeled guard substring (`requires_3x3` / `require_full_extent`) so an unrelated failure can no longer green the test. Addressed in d9eaaeb.

File: src/frontends/pytorch/src/op/svd.cpp [hunk 7]
```diff
@@ -0,0 +1,278 @@
...
+#include "utils.hpp"
+
+namespace ov {
+namespace frontend {
+namespace pytorch {
+namespace op {
+
+using namespace ov::op;
+
+namespace {
```
[the remark this answers is not in the document]
Comment (evkotov) [reply src/frontends/pytorch/src/op/svd.cpp:33]: Kept as-is intentionally: `using namespace ov::op;` at namespace scope is the established convention in this directory — 149 of 178 `src/frontends/pytorch/src/op/*.cpp` files open with exactly this line (e.g. `softmax.cpp`, `topk.cpp`, `argmax_argmin.cpp`). The new `svd.cpp` / `linalg_det.cpp` translators follow their neighbours. The project guideline forbids `using namespace` in **headers** and at **global** scope; these are `.cpp` translators matching local house style. Happy to switch to explicit qualification if you'd prefer, but that would make the two new files inconsistent with the rest of the directory.

File: src/frontends/pytorch/src/op/max_poolnd.cpp [hunk 8]
```diff
@@ -28,9 +31,181 @@ namespace pytorch {
...
+    return guarded;
+}
+
+// Decompose a max_pool whose kernel_size is only known at runtime. OpenVINO's MaxPool takes the
+// window as constructor attributes, so a runtime-computed kernel (e.g. F.max_pool2d(x, [1,
+// x.size(3)])) is only handled as a "global" pool over the full extent of a spatial axis -- exactly
+// a ReduceMax (keep_dims=True to match MaxPool). Any other configuration is rejected.
+// Per spatial axis, elem_is_const/elem_const_val classify the kernel element: constant 1 = window
+// of 1 (identity), non-constant = full-extent pool (reduce, via negative axes for any rank/batch).
+OutputVector translate_max_pool_dynamic_kernel(const NodeContext& context,
```
Comment (mvafin) [inline src/frontends/pytorch/src/op/max_poolnd.cpp:82]: In SAM-6D the kernel is not really dynamic. It is static. PyTorch FE makes it dynamic because it assumes all shapes to be dynamic. To help FE with shape resolution, use `input` argument with specified desired shape when calling `convert_model`. If done that way very likely dynamic kernel max_pool will be not needed.

**Review by mvafin — CHANGES_REQUESTED** [review]
Dynamic kernel max_pool is very restricted, it only supports narrow subset of inputs and in fact not even required for SAM-6D if user specifies shape. Check if shape provided by user will solve the issue and remove max_pool change if it will.
